bootstrap/bootstrap.go

```diff
@@ -122,9 +122,18 @@ func RunAndReturnWaitGroup(
 			fatalError(fmt.Errorf("failed to create SecretProvider: %s", err.Error()), lc)
 		}
 
-		// Bypass the zero trust zitidfied transport for Core Keeper Configuration client
-		// Should leverage the HttpTransportFromService function from zerotrust pkg, set the default transport for now
-		secretProvider.SetHttpTransport(http.DefaultTransport)
+		serviceInfo := serviceConfig.GetBootstrap().Service
+		if serviceInfo != nil {
+			if serviceInfo.SecurityOptions[config.SecurityModeKey] == zerotrust.ZeroTrustMode {
```

Comment:
Sorry, yes you're right, this is higher in the chain. I still don't understand what's happening on your side and what's not working to make you need this change. 

With this change at this point, I continue to see the zero trust client refused to be set:
```
level=WARN ts=2024-07-22T20:26:37.012965852Z app=core-metadata source=secure.go:500 msg="refusing to override httpRoundTripper, already set"
```

With the `// bypass` removed, I no longer see any "refusing to override" messages.

Can you show me what the error was that made you put this in place? Can I test/verify? I put a pr up that's currently on HOLD to undo these changes. 

I would like to understand what the problem was that made you go down this path. Thanks!

Reply:
Sure, and I haven't fully understand the zero trust structure yet so please correct me if I'm wrong.

There was a new service added after you integrated openZiti into EdgeX which is called `edgex-core-keeper`.
Core Keeper is a lightweight configuration and registry service that is aimed to replace Consul in the EdgeX architecture.

Every service has both configuration and registry clients in order to communicate with Consul or Keeper.
To my understanding, the zero trust structure does not cover these two HTTP clients since there is already an access token mechanism with Consul.

At that time, I was trying to avoid the nil pointer dereference error [here](https://github.com/edgexfoundry/go-mod-core-contracts/blob/main/clients/http/utils/common.go#L66) first, but end up having the side-effect you have reported.

In order to fit in to the zero trust structure, we need to set the HTTP transport for both these HTTP clients as well.

As for the "refusing to override" messages, we could probably fix it with something like:
```go
if (bootstrapConfig.Registry.Type == "keeper") {
    // enable zero trust and set the HTTP transport based on the service info with zerotrust.HttpTransportFromService()
}
```

Reply:
What was nil on that line? authInjector or authInjector.RoundTripper? FWIW the PR I put up does exactly what you describe with respect to a "keeper" workaround but i'd only consider it a stop-gap solution. (see https://github.com/edgexfoundry/go-mod-bootstrap/pull/735/files)

> In order to fit into the zero trust structure, we need to set the HTTP transport for both these HTTP clients as well.

#### How OpenZIti SHOULD work....

When zero trust mode is enabled, OpenZiti ends up getting configured with a "fallback" dialer. When a network connection is required, if there is no OpenZiti service found (either it doesn't exist or the identity isn't authorized) then OpenZiti will use the fallback to make a connection. If you look at the existing code and trace it, you'll see that ends up using the default net dialer. Right now there was no support for a fallback other than this since it was never needed... 

#### New PR to your branch

I have added this support on a PR to your issue-734 branch. find that pr here: https://github.com/jackchenjc/go-mod-bootstrap/pull/1

Reply:
Sorry, I got it. That is my bad. I was confused at that time.

The `authInjector` is the one which is nil in `core-common-configuration` at the time I was testing Keeper.
So the nil issue didn't come from go-mod-bootstrap or any OpenZiti related changes and should be fixed in `core-common-configuration` which I have addressed here (https://github.com/edgexfoundry/edgex-go/blob/main/internal/core/common_config/main.go#L91) and you don't even have to know about it.

For now, I think we can remove all these bypass statements and everything should work as expected as it used to be.
As for running Core Keeper with zero trust mode enabled, we will make more tests with edgex-compose builder and hope to get your assistance then if we need any help.

I will take a look at your PR on my branch and see if we need that to work with Keeper. Many thanks.

Reply:
Great. Thanks. As for that error, it's not actually an error and it snuck in recently as OpenZiti is moving towards distrbuted controllers and is safe to ignore. I'm sure at some point an update will clean that up and remove it:

> level=ERROR ts=2024-07-22T12:11:19.769044503Z app=device-virtual source=logging.go:35 msg="openziti: error listing controllers, continuing with 1 default configured controller

>As for running Core Keeper with zero trust mode enabled, we will make more tests with edgex-compose builder and hope to get your assistance then if we need any help.

Yea please let me know when you run into a problem with it. The PR I put up is probably useful on the whole anyway, so it's worth taking a look at and understanding (and incoporating eventually). I'm happy to talk to you or help with core keeper when you're ready. Including enabling it for zero trust... Just let me know and if we need to get on a call late my time (or late yours, whatever :) ) and we can work through it